File: leetcode/0328/0328.js
var oddEvenList = function (head) {
  if (head == null) {
    return head;
  }
  let odd = head;         // 扫描奇数结点
  let even = head.next;   // 扫描偶数结点
  let evenHead = even;    // 保存偶链的头结点

  while (even != null && even.next != null) { 
    odd.next = even.next; // even.next是下一个奇数结点
    odd = odd.next;       // odd 推进到下一个奇数结点
    even.next = odd.next; // 下一个奇数结点的next是下一个偶数结点
    even = even.next;     // even 推进到下一个偶数结点
  }
  odd.next = evenHead;    // 奇链连上偶链
  return head;
};

//作者：xiao_ben_zhu
//链接：https://leetcode-cn.com/problems/odd-even-linked-list/solution/shou-hua-tu-jie-328qi-ou-lian-biao-odd-even-linked/

File: leetcode/0328/0328.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (!head) return head;
        ListNode* evenHead = head->next;        //奇头是head，偶头就是head->next
        ListNode* odd = head, *even = evenHead; //分别定义奇偶尾指针

        while (even && even->next)
        {
            odd = odd->next = even->next;       //等于 odd->next = even->next; odd = odd->next;
            even = even->next = odd->next;
        }

        odd->next = evenHead;

        return head;
    }
};
